### PR TITLE
Crimp implementation

### PR DESCRIPTION
**Pull request by JoeARO**
Implementations of the Crimp Ruby Gem located here: https://github.com/bbc-news/crimp

This PR has almost all of the functionality that the Crimp Gem has however the Nested Hashes/Maps functionality isn't 100% there as of yet due to the Notation function treating the Maps as separate entities meaning we receive an incorrect signature.
This functionality isn't needed ASAP for our team however we should look at fixing this as soon as we can.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,93 @@
...
+    {:crimpex, "~> 0.1.0"}
+  ]
+end
+```
+
+Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
+and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
+be found at [https://hexdocs.pm/crimpex](https://hexdocs.pm/crimpex).
+
+# El-Crimp
```
Comment (ettomatic) [inline README.md]: Crimpex?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,93 @@
...
+Notation.notate([ b: 3 ])
+"3NbSAH"
+```
+
+## Changelog
+
+### 1.0.0-0
+
+Initial release.
+
```
Comment (ettomatic) [inline README.md:84]: I think you should mention here or somewhere what's still missing: eg failing tests, edge cases, etc

Comment (JoeARO) [reply README.md:84]: Now doing that actually thanks though :)

File: README.md (hunk 1 above)
Comment (JoeARO) [reply README.md]: 🤦‍♂ I did a search and replace for _ not -

Comment (JoeARO) [reply README.md]: Done!

File: README.md (hunk 2 above)
Comment (JoeARO) [reply README.md:84]: Have added it just below this, have also added that the project isn't currently on Hex!

File: README.md [hunk 3]
```diff
@@ -90,4 +88,10 @@ For testing we use ExUnit built into Elixir itself
 
 ```sh
 mix test
-```
\ No newline at end of file
+```
+
+## Caveats
+
+Currently this implementation of Crimp does not cover the functionallity that the original project intended, there is an issue with the Notation of Nested Hashes/Maps meaning they get sorted seperately causing the Signature to be incorrect. We are working to fix this and hopefully will update the project soon with that functionality.
```
Comment (ettomatic) [inline README.md:95]: great, I'd be super clear and add an example of what's not working compared to Ruby

Comment (JoeARO) [reply README.md:95]: Good thinking

Comment (JoeARO) [reply README.md:95]: Done!

File: lib/notation.ex [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  def notate(:true) do
+    "trueB"
+  end
+
+  def notate(nil) do
+    "_"
+  end
+
+  def notate(data) do
```
Comment (james-bowers) [inline lib/notation.ex]: is it possible to have multiple notate functions instead of this, which have guards to determine the type of data. E.g:

`def notate(data) when is_list(data)`
`def notate(data) when is_map(data)`
`def notate(data) when is_binary(data)`

File: lib/notation.ex [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  
+  defp sort_map(map) do
+    map
+    |> tuple()
+    |> Enum.sort(&string_compare/2)
+    |> Enum.map(&notate/1)
+    |> to_string
+  end
+
+  defp tuple(map) do
```
Comment (james-bowers) [inline lib/notation.ex:33]: would this work?
```
defp tuple(map) do
    map
    |> Enum.map(fn {k, v} ->
      [to_string(k), v]
    end)
end
```

File: lib/notation.ex [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  def string_compare(a, b) do
+    compare_objects(a) <= compare_objects(b)
+  end
+
+  def compare_objects(obj) do
+    case Util.typeof(obj) do
+      "list" -> Enum.sort(obj, &string_compare/2)
+      "map"  -> Enum.sort(tuple(obj), &string_compare/2)
+      ""
```
Comment (james-bowers) [inline lib/notation.ex]:
```suggestion
```

File: test/notation_test.exs [hunk 7]
```diff
@@ -0,0 +1,91 @@
+
```
Comment (james-bowers) [inline test/notation_test.exs]:
```suggestion
```

File: lib/notation.ex (hunk 5 above)
Comment (JoeARO) [reply lib/notation.ex:33]: ill give it a go thanks

File: lib/notation.ex (hunk 6 above)
Comment (JoeARO) [reply lib/notation.ex]: Already removed thanks, I think it was part way through mine and Ryans brainstorming for how to fix nested hash

**Review by ettomatic — APPROVED** [review]
good first iteration, I'd merge and then look at the edge cases.

**Review by james-bowers — APPROVED** [review]
